joint_space_mdp.py

```diff
@@ -0,0 +1,482 @@
+
+        end = time.time()
+        print("Total Runtime of Eval Policy: %.3f" % (end-start))
+
+    def update_policy(self):
+        """Use current value function to estimate new policy.
+        """
+        print("Updating Policy...")
+        num_states = float(len(self.valid_states))
+        total_time = 0
+
+        # for each state, find best action(s) to take
+        for (x, y) in self.valid_states:
+            start = time.time()
+            self.env.change_bottle_pos(
+                new_pos=[x, y, 0.1],
+                target_type=self.target_type)
+            (xi, yi) = self.state_to_idx((x, y))
+            best_value = 0
+            best_actions = []
+            sim_log = []
+            for ai, action in enumerate(self.A):
+                # if desired x,y,z is out of reach, skip this action
+                # validity depends on contact height
+                target_dist = np.linalg.norm(np.array([x, y]))
+                if (target_dist > self.env.arm.calc_max_horiz_dist(
+                        action.height)):
+                    continue
+
+                value, ns = self.env.run_sim_stochastic(action)
+
+                try:
+                    (nxi, nyi) = self.state_to_idx(ns)
+                    expected_future_value = self.V[nyi, nxi]
+                except IndexError:
+                    # just use value at current state if next state is out of  bounds
+                    # expected_future_cost = self.V[yi, xi]
+                    if self.cost_based:
+                        expected_future_value = self.OUT_OF_BOUNDS_COST
+                    else:
+                        expected_future_value = -1 * self.OUT_OF_BOUNDS_COST
+                    # expected_future_cost = 0
+
+                total_value = value + self.gamma*expected_future_value
+                sim_log.append((ns, value, expected_future_value, total_value))
+
+                if self.cost_based:
+                    found_better = (total_value < best_value)
+                else:
+                    found_better = (total_value > best_value)
+                if len(best_actions) == 0 or found_better:
+                    best_value = total_value
+                    best_actions = [ai]
+                elif math.isclose(total_value, best_value, abs_tol=1e-6):
+                    best_actions.append(ai)
+
+            # self.plot_sim_results(sim_log)
+            if self.DEBUG:
+                filename = "logs/action_costs_of_%d_%d_iter_%d" % (
+                    xi, yi, self.main_iter)
+                np.save(filename, sim_log)
+
+            self.P[yi][xi] = best_actions
+            end = time.time()
+            print("Time(s) for one state: %.3f" % (end - start))
+            total_time += (end-start)
+        print("Total Runtime of Update Policy: %.3f" % total_time)
+
+    def init_action_space(self, run_full_mdp):
+        A = []  # need to maintain order
+        self.da = math.pi/80
+        if run_full_mdp:
```

Comment:
Not sure why this boolean would map to what's happening in the body of the if/then. Comment? (Btw I normally think of variable names and code logic as "what" and comments as "why" - so if the why is clear, no comments necessary, but in this case I find the why not very clear.)